More helpful message when command doesn't exist?

- [x] I'd be willing to implement this feature ([contributing guide](https://github.com/pnpm/pnpm/blob/main/CONTRIBUTING.md))

**Describe the user story**

When someone runs `pnpm run` with a non-existent command, the current error message is not very informative:

```plaintext
$ pnpm run buil
 ERR_PNPM_NO_SCRIPT  Missing script: buil
```

Similar with just `pnpm` and the command:

```plaintext
$ pnpm buil
 ERR_PNPM_RECURSIVE_EXEC_FIRST_FAIL  Command failed with ENOENT: buil
spawn buil ENOENT
```

Many users wouldn't know what `ENOENT` is. It'd be nice to have a more explicit message saying what went wrong.

**Describe the solution you'd like**

How about:
* Explicitly saying in plain text that the command is missing
* If a similar command exists, suggesting it

```plaintext
$ pnpm run buil
 ERR_PNPM_NO_SCRIPT  Missing script: buil
Command "buil" not found. Did you mean "pnpm run build"?
```

```plaintext
$ pnpm buil
 ERR_PNPM_RECURSIVE_EXEC_FIRST_FAIL  Command failed with ENOENT: buil
Command "buil" not found. Did you mean "pnpm run build"?
```

**Describe the drawbacks of your solution**

...it's less technically precise, I suppose? Feels very pure-positive as a change to me. 🤷 

**Describe alternatives you've considered**

Here's what npm outputs:

```plaintext
$ npm run buil
npm ERR! Missing script: "buil"
npm ERR! 
npm ERR! Did you mean this?
npm ERR!     npm run build # run the "build" package script
npm ERR! 
npm ERR! To see a list of scripts, run:
npm ERR!   npm run

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/josh/.npm/_logs/2023-05-01T19_10_14_940Z-debug-0.log
```

```plaintext
$ npm buil
Unknown command: "buil"

Did you mean this?
    npm run build # run the "build" package script

To see a list of supported npm commands, run:
  npm help
```

...and here's Yarn v1:

```plaintext
$ yarn run buil
yarn run v1.22.19
error Command "buil" not found. Did you mean "build"?
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

```plaintext
$ yarn buil
yarn run v1.22.19
error Command "buil" not found. Did you mean "build"?
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

--- a/.changeset/pink-tips-rule.md
+++ b/.changeset/pink-tips-rule.md
@@ -1,0 +1,6 @@
+---
+"@pnpm/plugin-commands-script-runners": minor
+"@pnpm/cli-utils": patch
+---
+
+feat: expanded missing command error, including 'did you mean'

--- a/.changeset/pink-tips-rule.md
+++ b/.changeset/pink-tips-rule.md
@@ -1,6 +1,7 @@
 ---
 "@pnpm/plugin-commands-script-runners": minor
 "@pnpm/cli-utils": patch
+"pnpm": patch
 ---
 
-feat: expanded missing command error, including 'did you mean'
+Expanded missing command error, including 'did you mean' [#6492](https://github.com/pnpm/pnpm/issues/6492).

--- a/__typings__/local.d.ts
+++ b/__typings__/local.d.ts
@@ -39,7 +39,7 @@ declare module '@pnpm/npm-package-arg' {
   export = anything
 }
 
-declare module '@zkochan/which' {
+declare module '@pnpm/which' {
   const anything: any
   export = anything
 }

--- a/cli/cli-utils/src/readProjectManifest.ts
+++ b/cli/cli-utils/src/readProjectManifest.ts
@@ -23,7 +23,7 @@ export async function readProjectManifestOnly (
   opts: {
     engineStrict?: boolean
     nodeVersion?: string
-  }
+  } = {}
 ): Promise<ProjectManifest> {
   const manifest = await utils.readProjectManifestOnly(projectDir)
   packageIsInstallable(projectDir, manifest as any, opts) // eslint-disable-line @typescript-eslint/no-explicit-any

--- a/exec/plugin-commands-script-runners/package.json
+++ b/exec/plugin-commands-script-runners/package.json
@@ -58,6 +58,7 @@
     "@pnpm/store-path": "workspace:*",
     "@pnpm/types": "workspace:*",
     "@zkochan/rimraf": "^2.1.2",
+    "didyoumean2": "^5.0.0",
     "execa": "npm:safe-execa@0.1.2",
     "p-limit": "^3.1.0",
     "path-exists": "^4.0.0",

--- a/exec/plugin-commands-script-runners/package.json
+++ b/exec/plugin-commands-script-runners/package.json
@@ -67,7 +67,7 @@
     "ramda": "npm:@pnpm/ramda@0.28.1",
     "realpath-missing": "^1.1.0",
     "render-help": "^1.0.3",
-    "which": "^3.0.0",
+    "which": "npm:@zkochan/which@^2.0.3",
     "write-json-file": "^4.3.0"
   },
   "peerDependencies": {

--- a/exec/plugin-commands-script-runners/package.json
+++ b/exec/plugin-commands-script-runners/package.json
@@ -67,7 +67,7 @@
     "ramda": "npm:@pnpm/ramda@0.28.1",
     "realpath-missing": "^1.1.0",
     "render-help": "^1.0.3",
-    "which": "npm:@zkochan/which@^2.0.3",
+    "which": "npm:@pnpm/which@^3.0.1",
     "write-json-file": "^4.3.0"
   },
   "peerDependencies": {

--- a/exec/plugin-commands-script-runners/package.json
+++ b/exec/plugin-commands-script-runners/package.json
@@ -40,6 +40,7 @@
     "@pnpm/registry-mock": "3.8.0",
     "@types/is-windows": "^1.0.0",
     "@types/ramda": "0.28.20",
+    "@types/which": "^2.0.2",
     "is-windows": "^1.0.2",
     "write-yaml-file": "^5.0.0"
   },
@@ -66,6 +67,7 @@
     "ramda": "npm:@pnpm/ramda@0.28.1",
     "realpath-missing": "^1.1.0",
     "render-help": "^1.0.3",
+    "which": "^3.0.0",
     "write-json-file": "^4.3.0"
   },
   "peerDependencies": {

--- a/exec/plugin-commands-script-runners/src/buildCommandNotFoundHint.ts
+++ b/exec/plugin-commands-script-runners/src/buildCommandNotFoundHint.ts
@@ -9,7 +9,7 @@ export function buildCommandNotFoundHint (scriptName: string, scripts?: PackageS
   })
 
   if (nearestCommand) {
-    hint += ` Did you mean "pnpm run ${nearestCommand}"`
+    hint += ` Did you mean "pnpm run ${nearestCommand}"?`
   }
 
   return hint

--- a/exec/plugin-commands-script-runners/src/buildCommandNotFoundHint.ts
+++ b/exec/plugin-commands-script-runners/src/buildCommandNotFoundHint.ts
@@ -1,0 +1,16 @@
+import { type PackageScripts } from '@pnpm/types'
+import didYouMean, { ReturnTypeEnums } from 'didyoumean2'
+
+export function buildCommandNotFoundHint (scriptName: string, scripts?: PackageScripts | undefined) {
+  let hint = `Command "${scriptName}" not found.`
+
+  const nearestCommand = scripts && didYouMean(scriptName, Object.keys(scripts), {
+    returnType: ReturnTypeEnums.FIRST_CLOSEST_MATCH,
+  })
+
+  if (nearestCommand) {
+    hint += ` Did you mean "pnpm run ${nearestCommand}"`
+  }
+
+  return hint
+}

--- a/exec/plugin-commands-script-runners/src/exec.ts
+++ b/exec/plugin-commands-script-runners/src/exec.ts
@@ -19,6 +19,7 @@ import {
   shorthands as runShorthands,
 } from './run'
 import { PnpmError } from '@pnpm/error'
+import which from 'which'
 import writeJsonFile from 'write-json-file'
 import { buildCommandNotFoundHint } from './buildCommandNotFoundHint'
 
@@ -209,7 +210,9 @@ export async function handler (
           result[prefix].status = 'passed'
           result[prefix].duration = getExecutionDuration(startTime)
         } catch (err: any) { // eslint-disable-line
-          if (!opts.recursive && typeof err.exitCode === 'number') {
+          if (await isErrorCommandNotFound(params[0], err)) {
+            err.hint = buildCommandNotFoundHint(params[0], (await readProjectManifestOnly(opts.dir)).scripts)
+          } else if (!opts.recursive && typeof err.exitCode === 'number') {
             exitCode = err.exitCode
             return
           }
@@ -225,10 +228,6 @@ export async function handler (
 
           if (!opts.bail) {
             return
-          }
-
-          if (err.originalMessage === `spawn ${params[0]} ENOENT`) {
-            err.hint = buildCommandNotFoundHint(params[0], (await readProjectManifestOnly(opts.dir)).scripts)
           }
 
           if (!err['code']?.startsWith('ERR_PNPM_')) {
@@ -252,4 +251,20 @@ export async function handler (
   })
   throwOnCommandFail('pnpm recursive exec', result)
   return { exitCode }
+}
+
+interface CommandError extends Error {
+  originalMessage: string
+  shortMessage: string
+}
+
+async function isErrorCommandNotFound (command: string, error: CommandError) {
+  // Mac/Linux
+  if (error.originalMessage === `spawn ${command} ENOENT`) {
+    return true
+  }
+
+  // Windows
+  return error.shortMessage === `Command failed with exit code 1: ${command}` &&
+    !(await which(command, { nothrow: true }))
 }

--- a/exec/plugin-commands-script-runners/src/exec.ts
+++ b/exec/plugin-commands-script-runners/src/exec.ts
@@ -1,5 +1,5 @@
 import path from 'path'
-import { docsUrl, type RecursiveSummary, throwOnCommandFail } from '@pnpm/cli-utils'
+import { docsUrl, type RecursiveSummary, throwOnCommandFail, readProjectManifestOnly } from '@pnpm/cli-utils'
 import { type Config, types } from '@pnpm/config'
 import { makeNodeRequireOption } from '@pnpm/lifecycle'
 import { logger } from '@pnpm/logger'
@@ -20,6 +20,7 @@ import {
 } from './run'
 import { PnpmError } from '@pnpm/error'
 import writeJsonFile from 'write-json-file'
+import { buildCommandNotFoundHint } from './buildCommandNotFoundHint'
 
 export const shorthands = {
   parallel: runShorthands.parallel,
@@ -224,6 +225,10 @@ export async function handler (
 
           if (!opts.bail) {
             return
+          }
+
+          if (err.originalMessage === `spawn ${params[0]} ENOENT`) {
+            err.hint = buildCommandNotFoundHint(params[0], (await readProjectManifestOnly(opts.dir)).scripts)
           }
 
           if (!err['code']?.startsWith('ERR_PNPM_')) {

--- a/exec/plugin-commands-script-runners/src/exec.ts
+++ b/exec/plugin-commands-script-runners/src/exec.ts
@@ -212,7 +212,8 @@ export async function handler (
         } catch (err: any) { // eslint-disable-line
           if (await isErrorCommandNotFound(params[0], err)) {
             err.hint = buildCommandNotFoundHint(params[0], (await readProjectManifestOnly(opts.dir)).scripts)
-          } else if (!opts.recursive && typeof err.exitCode === 'number') {
+          }
+          if (!opts.recursive && typeof err.exitCode === 'number') {
             exitCode = err.exitCode
             return
           }

--- a/exec/plugin-commands-script-runners/src/exec.ts
+++ b/exec/plugin-commands-script-runners/src/exec.ts
@@ -212,8 +212,7 @@ export async function handler (
         } catch (err: any) { // eslint-disable-line
           if (await isErrorCommandNotFound(params[0], err)) {
             err.hint = buildCommandNotFoundHint(params[0], (await readProjectManifestOnly(opts.dir)).scripts)
-          }
-          if (!opts.recursive && typeof err.exitCode === 'number') {
+          } else if (!opts.recursive && typeof err.exitCode === 'number') {
             exitCode = err.exitCode
             return
           }

--- a/exec/plugin-commands-script-runners/src/run.ts
+++ b/exec/plugin-commands-script-runners/src/run.ts
@@ -21,6 +21,7 @@ import renderHelp from 'render-help'
 import { runRecursive, type RecursiveRunOpts, getSpecifiedScripts as getSpecifiedScriptWithoutStartCommand } from './runRecursive'
 import { existsInDir } from './existsInDir'
 import { handler as exec } from './exec'
+import { buildCommandNotFoundHint } from './buildCommandNotFoundHint'
 
 export const IF_PRESENT_OPTION = {
   'if-present': Boolean,
@@ -197,7 +198,10 @@ so you may run "pnpm -w run ${scriptName}"`,
         })
       }
     }
-    throw new PnpmError('NO_SCRIPT', `Missing script: ${scriptName}`)
+
+    throw new PnpmError('NO_SCRIPT', `Missing script: ${scriptName}`, {
+      hint: buildCommandNotFoundHint(scriptName, manifest.scripts),
+    })
   }
   const lifecycleOpts: RunLifecycleHookOptions = {
     depPath: dir,

--- a/exec/plugin-commands-script-runners/test/exec.e2e.ts
+++ b/exec/plugin-commands-script-runners/test/exec.e2e.ts
@@ -832,5 +832,5 @@ test('pnpm exec command not found', async () => {
   } catch (err: any) { // eslint-disable-line
     error = err
   }
-  expect(error?.hint).toBe('Command "buil" not found. Did you mean "pnpm run build"')
+  expect(error?.hint).toBe('Command "buil" not found. Did you mean "pnpm run build"?')
 })

--- a/exec/plugin-commands-script-runners/test/exec.e2e.ts
+++ b/exec/plugin-commands-script-runners/test/exec.e2e.ts
@@ -811,3 +811,26 @@ test('pnpm recursive exec report summary with --bail', async () => {
   expect(executionStatus[path.resolve('project-3')].status).toBe('running')
   expect(executionStatus[path.resolve('project-4')].status).toBe('queued')
 })
+
+test('pnpm exec command not found', async () => {
+  prepare({
+    scripts: {
+      build: 'echo hello',
+    },
+  })
+
+  const { selectedProjectsGraph } = await readProjects(process.cwd(), [])
+  let error!: Error & { hint: string }
+  try {
+    await exec.handler({
+      ...DEFAULT_OPTS,
+      dir: process.cwd(),
+      recursive: false,
+      bail: true,
+      selectedProjectsGraph,
+    }, ['buil'])
+  } catch (err: any) { // eslint-disable-line
+    error = err
+  }
+  expect(error?.hint).toBe('Command "buil" not found. Did you mean "pnpm run build"')
+})

--- a/exec/plugin-commands-script-runners/test/index.ts
+++ b/exec/plugin-commands-script-runners/test/index.ts
@@ -585,3 +585,22 @@ test('pnpm run with RegExp script selector with flag should throw error', async 
   }
   expect(err.message).toBe('RegExp flags are not supported in script command selector')
 })
+
+test('pnpm run with slightly incorrect command suggests correct one', async () => {
+  prepare({
+    scripts: {
+      build: 'echo 0',
+    },
+  })
+
+  await expect(run.handler({
+    dir: process.cwd(),
+    extraBinPaths: [],
+    extraEnv: {},
+    rawConfig: {},
+    workspaceConcurrency: 1,
+  }, ['buil'])).rejects.toEqual(expect.objectContaining({
+    code: 'ERR_PNPM_NO_SCRIPT',
+    hint: 'Command "buil" not found. Did you mean "pnpm run build"',
+  }))
+})

--- a/exec/plugin-commands-script-runners/test/index.ts
+++ b/exec/plugin-commands-script-runners/test/index.ts
@@ -601,6 +601,6 @@ test('pnpm run with slightly incorrect command suggests correct one', async () =
     workspaceConcurrency: 1,
   }, ['buil'])).rejects.toEqual(expect.objectContaining({
     code: 'ERR_PNPM_NO_SCRIPT',
-    hint: 'Command "buil" not found. Did you mean "pnpm run build"',
+    hint: 'Command "buil" not found. Did you mean "pnpm run build"?',
   }))
 })

--- a/pkg-manager/plugin-commands-installation/package.json
+++ b/pkg-manager/plugin-commands-installation/package.json
@@ -90,7 +90,7 @@
     "@yarnpkg/parsers": "3.0.0-rc.42",
     "@zkochan/rimraf": "^2.1.2",
     "@zkochan/table": "^2.0.1",
-    "@zkochan/which": "^2.0.3",
+    "@pnpm/which": "^3.0.1",
     "chalk": "^4.1.2",
     "ci-info": "^3.8.0",
     "enquirer": "^2.3.6",

--- a/pkg-manager/plugin-commands-installation/src/nodeExecPath.ts
+++ b/pkg-manager/plugin-commands-installation/src/nodeExecPath.ts
@@ -1,5 +1,5 @@
 import { promises as fs } from 'fs'
-import which from '@zkochan/which'
+import which from '@pnpm/which'
 
 export async function getNodeExecPath () {
   try {

